pkg/sql/catalog/descs/temporary_descriptors.go

```diff
@@ -24,19 +24,57 @@ import (
 )
 
 type temporaryDescriptors struct {
-	codec       keys.SQLCodec
-	sessionData *sessiondata.SessionData
+	codec                   keys.SQLCodec
+	temporarySchemaProvider TemporarySchemaProvider
 }
 
 func makeTemporaryDescriptors(
-	codec keys.SQLCodec, data *sessiondata.SessionData,
+	codec keys.SQLCodec, temporarySchemaProvider TemporarySchemaProvider,
 ) temporaryDescriptors {
 	return temporaryDescriptors{
-		codec:       codec,
-		sessionData: data,
+		codec:                   codec,
+		temporarySchemaProvider: temporarySchemaProvider,
 	}
 }
 
+// TemporarySchemaProvider is an interface that provides temporary schema
+// details on the current session.
+type TemporarySchemaProvider interface {
+	GetTemporarySchemaName() string
+	GetTemporarySchemaIDForDB(descpb.ID) (descpb.ID, bool)
+	MaybeGetDatabaseForTemporarySchemaID(descpb.ID) (descpb.ID, bool)
+}
+
+type temporarySchemaProviderImpl struct {
```

Comment:
In what ways is this better than making the impl a `sessiondata.Stack` for its underlying type?

Reply:
can't import descpb.ID from there `sessiondata` :D

Reply:
Like why 

```
type temporarySchemaProviderImpl struct {
   sds *sessiondata.Stack
}
```

and not:

```
type temporarySchemaProviderImpl sessiondata.Stack
```

The latter let's you do everything that the former does and avoids an allocation.

Reply:
Imagine an sds() method that does the cast.

Reply:
we need to be a pointer since otherwise the collection in `extraTxnState.collectionFactory` object points to a copy of the stack, which is incorrect.

Reply:
I'm still not following. Consider:

```go
type temporarySchemaProviderImpl sessiondata.Stack

// NewTemporarySchemaProvider creates a TemporarySchemaProvider.
func NewTemporarySchemaProvider(sds *sessiondata.Stack) TemporarySchemaProvider {
	return (*temporarySchemaProviderImpl)(sds)
}

func (impl *temporarySchemaProviderImpl) sds() *sessiondata.Stack {
    return (*sessionadata.Stack)(impl)
}

// GetTemporarySchemaName implements the TemporarySchemaProvider interface.
func (impl *temporarySchemaProviderImpl) GetTemporarySchemaName() string {
	return impl.sds().Top().TemporarySchemaName
}

// GetTemporarySchemaIDForDB implements the TemporarySchemaProvider interface.
func (impl *temporarySchemaProviderImpl) GetTemporarySchemaIDForDB(id descpb.ID) (descpb.ID, bool) {
	ret, found := impl.sds().Top().GetTemporarySchemaIDForDB(uint32(id))
	return descpb.ID(ret), found
}

// MaybeGetDatabaseForTemporarySchemaID implements the TemporarySchemaProvider interface.
func (impl *temporarySchemaProviderImpl) MaybeGetDatabaseForTemporarySchemaID(
	id descpb.ID,
) (descpb.ID, bool) {
	ret, found := impl.sds().Top().MaybeGetDatabaseForTemporarySchemaID(uint32(id))
	return descpb.ID(ret), found
}
```

Reply:
ultimately it may make sense to move temp tables out of session data and make it a global state to avoid this stuff (it's not SET LOCAL affected), but that's a bit more than i'm intended to do right now.

Reply:
ah wait, i failed to understand `type temporarySchemaProviderImpl sessiondata.Stack`

Reply:
ok done, but had to do this for the casting to work:

```
 // GetTemporarySchemaName implements the TemporarySchemaProvider interface.
 func (impl *temporarySchemaProviderImpl) GetTemporarySchemaName() string {
       return (*sessiondata.Stack)(impl).Top().TemporarySchemaName
 }
```

sorry for the confusion